File: windows/runner/mihomo_core.cpp
// MihomoCore.cpp - Mihomo Core Manager Implementation for Windows
#include "mihomo_core.h"

#include <winhttp.h>
#include <shlwapi.h>
#include <fstream>
#include <sstream>
#include <regex>
#include <chrono>
#include <iostream>

#pragma comment(lib, "winhttp.lib")
#pragma comment(lib, "shlwapi.lib")

MihomoCore& MihomoCore::GetInstance() {
    static MihomoCore instance;
    return instance;
}

MihomoCore::MihomoCore()
    : controllerHost_("127.0.0.1"),
      controllerPort_(9090),
      processHandle_(nullptr),
      processThread_(nullptr),
      processId_(0),
      isRunning_(false),
      stopMonitoring_(false),
      lastUpload_(0),
      lastDownload_(0),
      lastTime_(0),
      state_("disconnected") {}

MihomoCore::~MihomoCore() {
    Stop();
}

bool MihomoCore::Init(const std::string& workDir) {
    workDir_ = workDir;

    // Ensure work directory exists
    CreateDirectoryA(workDir.c_str(), nullptr);

    // Core binary path
    corePath_ = workDir + "\\mihomo.exe";

    // Check if core exists
    if (GetFileAttributesA(corePath_.c_str()) == INVALID_FILE_ATTRIBUTES) {
        if (errorCallback_) {
            errorCallback_("Core binary not found: " + corePath_);
        }
        return false;
    }

    return true;
}

bool MihomoCore::Start(const std::string& configPath) {
    if (isRunning_) {
        return true;
    }

    configPath_ = configPath;
    ParseControllerSettings(configPath);

    // Build command line
    std::string cmdLine = "\"" + corePath_ + "\" -d \"" + workDir_ + "\" -f \"" + configPath + "\"";

    STARTUPINFOA si = {0};
    si.cb = sizeof(si);
    si.dwFlags = STARTF_USESHOWWINDOW;
    si.wShowWindow = SW_HIDE;

    PROCESS_INFORMATION pi = {0};

    // Create process
    if (!CreateProcessA(
            nullptr,
            const_cast<char*>(cmdLine.c_str()),
            nullptr,
            nullptr,
            FALSE,
            CREATE_NO_WINDOW,
            nullptr,
            workDir_.c_str(),
            &si,
            &pi)) {
        if (errorCallback_) {
            errorCallback_("Failed to start core process");
        }
        return false;
    }

    processHandle_ = pi.hProcess;
    processThread_ = pi.hThread;
    processId_ = pi.dwProcessId;

    // Wait a bit for core to start
    Sleep(500);

    // Check if process is still running
    DWORD exitCode;
    if (GetExitCodeProcess(processHandle_, &exitCode) && exitCode != STILL_ACTIVE) {
        if (errorCallback_) {
            errorCallback_("Core process exited immediately");
        }
        CloseHandle(processHandle_);
        CloseHandle(processThread_);
        processHandle_ = nullptr;
        processThread_ = nullptr;
        return false;
    }

    isRunning_ = true;
    state_ = "connected";
    stopMonitoring_ = false;

    if (stateCallback_) {
        stateCallback_(state_);
    }

    // Start monitoring
    StartTrafficMonitor();

    return true;
}

bool MihomoCore::Stop() {
    if (!isRunning_) {
        return true;
    }

    state_ = "disconnecting";
    if (stateCallback_) {
        stateCallback_(state_);
    }

    StopMonitoring();

    // Terminate process
    if (processHandle_) {
        TerminateProcess(processHandle_, 0);
        WaitForSingleObject(processHandle_, 3000);
        CloseHandle(processHandle_);
        processHandle_ = nullptr;
    }

    if (processThread_) {
        CloseHandle(processThread_);
        processThread_ = nullptr;
    }

    processId_ = 0;
    isRunning_ = false;
    state_ = "disconnected";

    if (stateCallback_) {
        stateCallback_(state_);
    }

    return true;
}

bool MihomoCore::ReloadConfig(const std::string& configPath) {
    std::string body = "{\"path\":\"" + configPath + "\"}";
    std::string response = HttpPut("/configs?force=true", body);
    if (!response.empty()) {
        configPath_ = configPath;
        ParseControllerSettings(configPath);
        return true;
    }
    return false;
}

bool MihomoCore::IsRunning() const {
    if (!isRunning_ || !processHandle_) {
        return false;
    }

    DWORD exitCode;
    if (GetExitCodeProcess(processHandle_, &exitCode)) {
        return exitCode == STILL_ACTIVE;
    }
    return false;
}

std::string MihomoCore::GetVersion() {
    std::string response = HttpGet("/version");
    if (!response.empty()) {
        // Simple JSON parsing for version
        std::regex versionRegex("\"version\"\\s*:\\s*\"([^\"]+)\"");
        std::smatch match;
        if (std::regex_search(response, match, versionRegex)) {
            return match[1].str();
        }
    }
    return "unknown";
}

MihomoCore::TrafficStats MihomoCore::GetTrafficStats() {
    TrafficStats stats = {0, 0, 0, 0};

    std::string response = HttpGet("/traffic");
    if (!response.empty()) {
        // Parse JSON
        std::regex upRegex("\"up\"\\s*:\\s*(\\d+)");
        std::regex downRegex("\"down\"\\s*:\\s*(\\d+)");
        std::smatch match;

        if (std::regex_search(response, match, upRegex)) {
            stats.upload = std::stoll(match[1].str());
        }
        if (std::regex_search(response, match, downRegex)) {
            stats.download = std::stoll(match[1].str());
        }
    }

    return stats;
}

int MihomoCore::TestDelay(const std::string& proxy, const std::string& url, int timeout) {
    std::string path = "/proxies/" + proxy + "/delay?timeout=" + std::to_string(timeout) + "&url=" + url;
    std::string response = HttpGet(path);

    if (!response.empty()) {
        std::regex delayRegex("\"delay\"\\s*:\\s*(\\d+)");
        std::smatch match;
        if (std::regex_search(response, match, delayRegex)) {
            return std::stoi(match[1].str());
        }
    }
    return -1;
}

bool MihomoCore::SwitchProxy(const std::string& selector, const std::string& proxy) {
    std::string body = "{\"name\":\"" + proxy + "\"}";
    std::string response = HttpPut("/proxies/" + selector, body);
    return !response.empty();
}

std::string MihomoCore::GetConnections() {
    return HttpGet("/connections");
}

std::string MihomoCore::GetLogs() {
    std::string logPath = workDir_ + "\\logs\\mihomo.log";
    std::ifstream file(logPath);
    if (file.is_open()) {
        std::stringstream buffer;
        buffer << file.rdbuf();
        return buffer.str();
    }
    return "No logs available";
}

std::string MihomoCore::ExportLogs() {
    std::string logs = GetLogs();
    auto now = std::chrono::system_clock::now();
    auto time = std::chrono::system_clock::to_time_t(now);

    std::string exportPath = workDir_ + "\\vortex_logs_" + std::to_string(time) + ".txt";
    std::ofstream file(exportPath);
    if (file.is_open()) {
        file << logs;
        file.close();
        return exportPath;
    }
    return "";
}

void MihomoCore::SetStateCallback(StateCallback callback) {
    stateCallback_ = callback;
}

void MihomoCore::SetTrafficCallback(TrafficCallback callback) {
    trafficCallback_ = callback;
}

void MihomoCore::SetLogCallback(LogCallback callback) {
    logCallback_ = callback;
}

void MihomoCore::SetErrorCallback(ErrorCallback callback) {
    errorCallback_ = callback;
}

void MihomoCore::ParseControllerSettings(const std::string& configPath) {
    std::ifstream file(configPath);
    if (!file.is_open()) return;

    std::stringstream buffer;
    buffer << file.rdbuf();
    std::string content = buffer.str();

    // Parse external-controller
    std::regex controllerRegex("external-controller:\\s*['\"]?([^'\":\\s]+):?(\\d+)?['\"]?");
    std::smatch match;
    if (std::regex_search(content, match, controllerRegex)) {
        controllerHost_ = match[1].str();
        if (!match[2].str().empty()) {
            controllerPort_ = std::stoi(match[2].str());
        }
    }

    // Parse secret
    std::regex secretRegex("secret:\\s*['\"]?([^'\"\\s]+)['\"]?");
    if (std::regex_search(content, match, secretRegex)) {
        controllerSecret_ = match[1].str();
    }
}

void MihomoCore::StartTrafficMonitor() {
    trafficThread_ = std::thread([this]() {
        while (!stopMonitoring_) {
            if (isRunning_ && trafficCallback_) {
                TrafficStats stats = GetTrafficStats();

                ULONGLONG now = GetTickCount64();
                double timeDelta = (now - lastTime_) / 1000.0;

                if (lastTime_ > 0 && timeDelta > 0) {
                    stats.uploadSpeed = static_cast<int64_t>((stats.upload - lastUpload_) / timeDelta);
                    stats.downloadSpeed = static_cast<int64_t>((stats.download - lastDownload_) / timeDelta);

                    trafficCallback_(stats);
                }

                lastUpload_ = stats.upload;
                lastDownload_ = stats.download;
                lastTime_ = now;
            }

            Sleep(1000);
        }
    });
}

void MihomoCore::StopMonitoring() {
    stopMonitoring_ = true;

    if (trafficThread_.joinable()) {
        trafficThread_.join();
    }
    if (logThread_.joinable()) {
        logThread_.join();
    }
}

std::string MihomoCore::HttpGet(const std::string& path) {
    HINTERNET hSession = nullptr;
    HINTERNET hConnect = nullptr;
    HINTERNET hRequest = nullptr;
    std::string result;

    try {
        hSession = WinHttpOpen(L"Vortex/1.0",
            WINHTTP_ACCESS_TYPE_DEFAULT_PROXY,
            WINHTTP_NO_PROXY_NAME,
            WINHTTP_NO_PROXY_BYPASS, 0);

        if (!hSession) throw std::runtime_error("WinHttpOpen failed");

        std::wstring wHost(controllerHost_.begin(), controllerHost_.end());
        hConnect = WinHttpConnect(hSession, wHost.c_str(), static_cast<INTERNET_PORT>(controllerPort_), 0);
        if (!hConnect) throw std::runtime_error("WinHttpConnect failed");

        std::wstring wPath(path.begin(), path.end());
        hRequest = WinHttpOpenRequest(hConnect, L"GET", wPath.c_str(),
            nullptr, WINHTTP_NO_REFERER, WINHTTP_DEFAULT_ACCEPT_TYPES, 0);
        if (!hRequest) throw std::runtime_error("WinHttpOpenRequest failed");

        // Add authorization header if secret is set
        if (!controllerSecret_.empty()) {
            std::wstring authHeader = L"Authorization: Bearer " +
                std::wstring(controllerSecret_.begin(), controllerSecret_.end());
            WinHttpAddRequestHeaders(hRequest, authHeader.c_str(), static_cast<DWORD>(-1),
                WINHTTP_ADDREQ_FLAG_ADD);
        }

        if (!WinHttpSendRequest(hRequest, WINHTTP_NO_ADDITIONAL_HEADERS, 0,
            WINHTTP_NO_REQUEST_DATA, 0, 0, 0)) {
            throw std::runtime_error("WinHttpSendRequest failed");
        }

        if (!WinHttpReceiveResponse(hRequest, nullptr)) {
            throw std::runtime_error("WinHttpReceiveResponse failed");
        }

        DWORD size = 0;
        DWORD downloaded = 0;
        do {
            if (!WinHttpQueryDataAvailable(hRequest, &size)) break;
            if (size == 0) break;

            std::vector<char> buffer(size + 1, 0);
            if (WinHttpReadData(hRequest, buffer.data(), size, &downloaded)) {
                result.append(buffer.data(), downloaded);
            }
        } while (size > 0);

    } catch (...) {
        // Ignore errors
    }

    if (hRequest) WinHttpCloseHandle(hRequest);
    if (hConnect) WinHttpCloseHandle(hConnect);
    if (hSession) WinHttpCloseHandle(hSession);

    return result;
}

std::string MihomoCore::HttpPut(const std::string& path, const std::string& body) {
    HINTERNET hSession = nullptr;
    HINTERNET hConnect = nullptr;
    HINTERNET hRequest = nullptr;
    std::string result;

    try {
        hSession = WinHttpOpen(L"Vortex/1.0",
            WINHTTP_ACCESS_TYPE_DEFAULT_PROXY,
            WINHTTP_NO_PROXY_NAME,
            WINHTTP_NO_PROXY_BYPASS, 0);

        if (!hSession) throw std::runtime_error("WinHttpOpen failed");

        std::wstring wHost(controllerHost_.begin(), controllerHost_.end());
        hConnect = WinHttpConnect(hSession, wHost.c_str(), static_cast<INTERNET_PORT>(controllerPort_), 0);
        if (!hConnect) throw std::runtime_error("WinHttpConnect failed");

        std::wstring wPath(path.begin(), path.end());
        hRequest = WinHttpOpenRequest(hConnect, L"PUT", wPath.c_str(),
            nullptr, WINHTTP_NO_REFERER, WINHTTP_DEFAULT_ACCEPT_TYPES, 0);
        if (!hRequest) throw std::runtime_error("WinHttpOpenRequest failed");

        // Add headers
        std::wstring headers = L"Content-Type: application/json\r\n";
        if (!controllerSecret_.empty()) {
            headers += L"Authorization: Bearer " +
                std::wstring(controllerSecret_.begin(), controllerSecret_.end()) + L"\r\n";
        }
        WinHttpAddRequestHeaders(hRequest, headers.c_str(), static_cast<DWORD>(-1), WINHTTP_ADDREQ_FLAG_ADD);

        if (!WinHttpSendRequest(hRequest, WINHTTP_NO_ADDITIONAL_HEADERS, 0,
            (LPVOID)body.c_str(), static_cast<DWORD>(body.length()), static_cast<DWORD>(body.length()), 0)) {
            throw std::runtime_error("WinHttpSendRequest failed");
        }

        if (!WinHttpReceiveResponse(hRequest, nullptr)) {
            throw std::runtime_error("WinHttpReceiveResponse failed");
        }

        // Check status code
        DWORD statusCode = 0;
        DWORD size = sizeof(statusCode);
        WinHttpQueryHeaders(hRequest, WINHTTP_QUERY_STATUS_CODE | WINHTTP_QUERY_FLAG_NUMBER,
            WINHTTP_HEADER_NAME_BY_INDEX, &statusCode, &size, WINHTTP_NO_HEADER_INDEX);

        if (statusCode >= 200 && statusCode < 300) {
            result = "success";
        }

    } catch (...) {
        // Ignore errors
    }

    if (hRequest) WinHttpCloseHandle(hRequest);
    if (hConnect) WinHttpCloseHandle(hConnect);
    if (hSession) WinHttpCloseHandle(hSession);

    return result;
}

File: macos/Runner/MihomoCore.swift
import Foundation
import Cocoa

/// Mihomo Core Manager for macOS
/// Handles core process lifecycle, REST API communication, and traffic statistics
class MihomoCore {
    static let shared = MihomoCore()

    struct TrafficStats {
        var upload: Int64 = 0
        var download: Int64 = 0
        var uploadSpeed: Int64 = 0
        var downloadSpeed: Int64 = 0
    }

    typealias StateCallback = (String) -> Void
    typealias TrafficCallback = (TrafficStats) -> Void
    typealias LogCallback = (String) -> Void
    typealias ErrorCallback = (String) -> Void

    private var workDir: String = ""
    private var corePath: String = ""
    private var configPath: String = ""
    private var state: String = "disconnected"

    private var controllerHost: String = "127.0.0.1"
    private var controllerPort: Int = 9090
    private var controllerSecret: String = ""

    private var process: Process?
    private var isRunning: Bool = false
    private var stopMonitoring: Bool = false

    private var trafficTimer: Timer?
    private var lastUpload: Int64 = 0
    private var lastDownload: Int64 = 0
    private var lastTime: Date?

    var stateCallback: StateCallback?
    var trafficCallback: TrafficCallback?
    var logCallback: LogCallback?
    var errorCallback: ErrorCallback?

    private init() {}

    /// Initialize core
    func initialize(workDir: String) -> Bool {
        self.workDir = workDir

        // Ensure work directory exists
        let fileManager = FileManager.default
        if !fileManager.fileExists(atPath: workDir) {
            do {
                try fileManager.createDirectory(atPath: workDir, withIntermediateDirectories: true)
            } catch {
                print("Failed to create work directory: \(error)")
                return false
            }
        }

        // Core binary path
        corePath = workDir + "/mihomo"

        // Check if core exists
        if !fileManager.fileExists(atPath: corePath) {
            // Try to extract from bundle
            if let bundlePath = Bundle.main.path(forResource: "mihomo", ofType: nil) {
                do {
                    try fileManager.copyItem(atPath: bundlePath, toPath: corePath)
                    // Make executable
                    try fileManager.setAttributes([.posixPermissions: 0o755], ofItemAtPath: corePath)
                } catch {
                    errorCallback?("Failed to extract core binary: \(error.localizedDescription)")
                    return false
                }
            } else {
                errorCallback?("Core binary not found")
                return false
            }
        }

        return true
    }

    /// Start core with config
    func start(configPath: String) -> Bool {
        if isRunning {
            return true
        }

        self.configPath = configPath
        parseControllerSettings(configPath: configPath)

        process = Process()
        process?.executableURL = URL(fileURLWithPath: corePath)
        process?.arguments = ["-d", workDir, "-f", configPath]
        process?.currentDirectoryURL = URL(fileURLWithPath: workDir)

        // Capture output
        let pipe = Pipe()
        process?.standardOutput = pipe
        process?.standardError = pipe

        pipe.fileHandleForReading.readabilityHandler = { [weak self] handle in
            let data = handle.availableData
            if let output = String(data: data, encoding: .utf8), !output.isEmpty {
                self?.logCallback?(output)
            }
        }

        do {
            try process?.run()

            // Wait a bit
            Thread.sleep(forTimeInterval: 0.5)

            if process?.isRunning == true {
                isRunning = true
                state = "connected"
                stopMonitoring = false
                stateCallback?(state)
                startTrafficMonitor()
                return true
            } else {
                errorCallback?("Core process exited immediately")
                return false
            }
        } catch {
            errorCallback?("Failed to start core: \(error.localizedDescription)")
            return false
        }
    }

    /// Stop core
    func stop() -> Bool {
        if !isRunning {
            return true
        }

        state = "disconnecting"
        stateCallback?(state)

        stopMonitoring = true
        trafficTimer?.invalidate()
        trafficTimer = nil

        process?.terminate()
        process?.waitUntilExit()
        process = nil

        isRunning = false
        state = "disconnected"
        stateCallback?(state)

        return true
    }

    /// Reload config
    func reloadConfig(configPath: String) -> Bool {
        let body = "{\"path\":\"\(configPath)\"}"
        let result = httpPut(path: "/configs?force=true", body: body)
        if result != nil {
            self.configPath = configPath
            parseControllerSettings(configPath: configPath)
            return true
        }
        return false
    }

    /// Check if running
    func checkIsRunning() -> Bool {
        return isRunning && (process?.isRunning ?? false)
    }

    /// Get state
    func getState() -> String {
        return state
    }

    /// Get version
    func getVersion() -> String {
        guard let response = httpGet(path: "/version") else {
            return "unknown"
        }

        if let data = response.data(using: .utf8),
           let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
           let version = json["version"] as? String {
            return version
        }
        return "unknown"
    }

    /// Get traffic stats
    func getTrafficStats() -> TrafficStats {
        var stats = TrafficStats()

        guard let response = httpGet(path: "/traffic") else {
            return stats
        }

        if let data = response.data(using: .utf8),
           let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
            stats.upload = json["up"] as? Int64 ?? 0
            stats.download = json["down"] as? Int64 ?? 0
        }

        return stats
    }

    /// Test proxy delay
    func testDelay(proxy: String, url: String, timeout: Int) -> Int {
        let encodedProxy = proxy.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? proxy
        let path = "/proxies/\(encodedProxy)/delay?timeout=\(timeout)&url=\(url)"

        guard let response = httpGet(path: path) else {
            return -1
        }

        if let data = response.data(using: .utf8),
           let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
           let delay = json["delay"] as? Int {
            return delay
        }
        return -1
    }

    /// Switch proxy
    func switchProxy(selector: String, proxy: String) -> Bool {
        let encodedSelector = selector.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? selector
        let body = "{\"name\":\"\(proxy)\"}"
        return httpPut(path: "/proxies/\(encodedSelector)", body: body) != nil
    }

    /// Get connections
    func getConnections() -> String {
        return httpGet(path: "/connections") ?? ""
    }

    /// Get logs
    func getLogs() -> String {
        let logPath = workDir + "/logs/mihomo.log"
        do {
            return try String(contentsOfFile: logPath, encoding: .utf8)
        } catch {
            return "No logs available"
        }
    }

    /// Export logs
    func exportLogs() -> String? {
        let logs = getLogs()
        let timestamp = Int(Date().timeIntervalSince1970)
        let exportPath = workDir + "/vortex_logs_\(timestamp).txt"

        do {
            try logs.write(toFile: exportPath, atomically: true, encoding: .utf8)
            return exportPath
        } catch {
            return nil
        }
    }

    // MARK: - Private Methods

    private func parseControllerSettings(configPath: String) {
        guard let content = try? String(contentsOfFile: configPath, encoding: .utf8) else {
            return
        }

        // Parse external-controller
        if let range = content.range(of: "external-controller:\\s*['\"]?([^'\":\\s]+):?(\\d+)?['\"]?",
                                      options: .regularExpression) {
            let match = String(content[range])
            let components = match.replacingOccurrences(of: "external-controller:", with: "")
                .trimmingCharacters(in: .whitespaces)
                .replacingOccurrences(of: "'", with: "")
                .replacingOccurrences(of: "\"", with: "")
                .split(separator: ":")

            if components.count >= 1 {
                controllerHost = String(components[0])
            }
            if components.count >= 2, let port = Int(components[1]) {
                controllerPort = port
            }
        }

        // Parse secret
        if let range = content.range(of: "secret:\\s*['\"]?([^'\"\\s]+)['\"]?",
                                      options: .regularExpression) {
            let match = String(content[range])
            controllerSecret = match.replacingOccurrences(of: "secret:", with: "")
                .trimmingCharacters(in: .whitespaces)
                .replacingOccurrences(of: "'", with: "")
                .replacingOccurrences(of: "\"", with: "")
        }
    }

    private func startTrafficMonitor() {
        trafficTimer = Timer.scheduledTimer(withTimeInterval: 1.0, repeats: true) { [weak self] _ in
            guard let self = self, self.isRunning, !self.stopMonitoring else { return }

            var stats = self.getTrafficStats()
            let now = Date()

            if let lastTime = self.lastTime {
                let timeDelta = now.timeIntervalSince(lastTime)
                if timeDelta > 0 {
                    stats.uploadSpeed = Int64(Double(stats.upload - self.lastUpload) / timeDelta)
                    stats.downloadSpeed = Int64(Double(stats.download - self.lastDownload) / timeDelta)
                    self.trafficCallback?(stats)
                }
            }

            self.lastUpload = stats.upload
            self.lastDownload = stats.download
            self.lastTime = now
        }
    }

    private func httpGet(path: String) -> String? {
        let urlString = "http://\(controllerHost):\(controllerPort)\(path)"
        guard let url = URL(string: urlString) else { return nil }

        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.timeoutInterval = 5

        if !controllerSecret.isEmpty {
            request.setValue("Bearer \(controllerSecret)", forHTTPHeaderField: "Authorization")
        }

        let semaphore = DispatchSemaphore(value: 0)
        var result: String?

        let task = URLSession.shared.dataTask(with: request) { data, response, error in
            if let data = data {
                result = String(data: data, encoding: .utf8)
            }
            semaphore.signal()
        }
        task.resume()
        semaphore.wait()

        return result
    }

    private func httpPut(path: String, body: String) -> String? {
        let urlString = "http://\(controllerHost):\(controllerPort)\(path)"
        guard let url = URL(string: urlString) else { return nil }

        var request = URLRequest(url: url)
        request.httpMethod = "PUT"
        request.httpBody = body.data(using: .utf8)
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.timeoutInterval = 5

        if !controllerSecret.isEmpty {
            request.setValue("Bearer \(controllerSecret)", forHTTPHeaderField: "Authorization")
        }

        let semaphore = DispatchSemaphore(value: 0)
        var result: String?

        let task = URLSession.shared.dataTask(with: request) { data, response, error in
            if let httpResponse = response as? HTTPURLResponse,
               httpResponse.statusCode >= 200 && httpResponse.statusCode < 300 {
                result = "success"
            }
            semaphore.signal()
        }
        task.resume()
        semaphore.wait()

        return result
    }
}
